File: solutions/20/linear_time_space_minimal_nesting.js
'use strict'; const log = console.log; (async ()=>{

const f = s => {
  if(typeof s !== 'string')
    throw new Error();

  const stack = [], pairs = new Map(
    [[')', '('], [']', '['], ['}', '{']]
  );

  for(const c of s) {
    if(!pairs.has(c)) {
      stack.push(c);
      continue;
    }

    const expectedPair = pairs.get(c),
      peekChar = stack.pop();
    if(expectedPair !== peekChar)
      return false;
  }

  return !stack.length;
};

const eq = require('assert').strictEqual;
eq(f('([)]'), false);
eq(f('{[]}'), true);
eq(f('()[]{}'), true);
eq(f('[[[['), false);

})();

File: solutions/20/linear_time_space_minimal_nesting.py
def f(s):
    from collections import deque
    if not s: return True
    if len(s) %  2: return False
    matcher = {')': '(', ']': '[', '}': '{'}
    stack = deque()

    for c in s:
        if c not in matcher:
            stack.append(c)
            continue

        expected_pair = matcher[c]
        if stack[-1] != expected_pair:
            return False

        stack.pop()

    return not stack

assert not f('([)]')
assert f('{[]}')
assert f('()[]{}')
assert not f('[[[[')
